packages/buckets/src/buckets.spec.ts

```diff
@@ -271,6 +273,107 @@ describe('Buckets...', function () {
+      if (isBrowser) return this.skip()
+      const { root } = await client.getOrCreate('pushes')
+
+      // Grab a reference to a relatively large file
+      const pth = path.join(__dirname, '../../..', 'testdata')
+      fileSize = fs.statSync(path.join(pth, 'file1.jpg')).size
+      const stream1 = fs.createReadStream(path.join(pth, 'file1.jpg'))
+      const rootKey = root?.key || ''
+
+      // Now create a relatively small stream that should finish first
+      const stream2 = new Duplex()
+      stream2.push(Buffer.from('some small amount of text'))
+      stream2.push(Buffer.from('that spans two buffers'))
+      stream2.push(null)
+
+      // Run them in series
+      await Promise.resolve()
+        .then(() => client.pushPath(rootKey, 'dir1/file1.jpg', stream1, { root }))
+        .then(() => client.pushPath(rootKey, 'dir1/file2.txt', stream2, { root }))
+        .then(() => {
+          throw new Error('wrong error')
+        }) // This should never fire
+        // Because we should jump down to this first
+        .catch((err) => expect(err.message).to.include('non-fast-forward'))
+
+      // Confirm that our second-level dir only includes one of them
+      const { item } = await client.listPath(rootKey, '')
+      expect(item?.items[1].items).to.have.length(1)
+    })
+
+    it.skip('should not be able to push in parallel', async function () {
+      //TODO: Right now, the Go client doesn't "locK" a bucket for pushes, this is something
+      // we should add/support to ensure buckets pushed in parallel don't corrupt things
+      if (isBrowser) return this.skip()
+      const { root } = await client.getOrCreate('pushes')
+
+      // Grab a reference to a relatively large file
+      const pth = path.join(__dirname, '../../..', 'testdata')
+      fileSize = fs.statSync(path.join(pth, 'file1.jpg')).size
+      const stream1 = fs.createReadStream(path.join(pth, 'file1.jpg'))
+      const rootKey = root?.key || ''
+
+      // Now create a relatively small stream that should finish first
+      const stream2 = new Duplex()
+      stream2.push(Buffer.from('some small amount of text'))
+      stream2.push(Buffer.from('that spans two buffers'))
+      stream2.push(null)
+
+      try {
+        await Promise.all([
+          // Note the names are different from above test
+          client.pushPath(rootKey, 'dir1/file3.jpg', stream1),
+          client.pushPath(rootKey, 'dir1/file4.txt', stream2),
+        ])
+        throw new Error('wrong error')
+      } catch (err) {
+        expect(err.message).to.include('non-fast-forward')
+      }
+      // Confirm that our second-level dir only includes one of them
+      const { item } = await client.listPath(rootKey, '')
+      expect(item?.items[1].items).to.have.length(1)
+    })
+  })
+
+  describe('aborting', function () {
+    it('should allow an abort controler to signal a cancel event on a push', async function () {
```

Comment:
Just making sure we can abort an inflight push. Here, we use an infinite stream of "bytes", that in theory would never close the connection with our remote bucket API, but that is also a little bit slow. If the abort thing works as expected, it'll kill the connection and when we try to list path afterwards, we should get an error. If it doesn't work, this test will hang and timeout...

Reply:
There actually is no reason to not test this in a browser, I might remove that check actually...